Remove some SciPy dependencies

tmarkovich opened the pull request:
I have removed the scipy dependencies by using the gamma function that the python math module provides, and implementing my own $\chi^2$ survival function. This was done so that I could use goftests with pypy.

I have also included a test for my new $\chi^2$ survival function, and verified that it agrees with the scipy provided version to machine precision.

tmarkovich committed:
Moved to using my home-grown chi^2 cdf

I see a 1e-15 match with scipy for a range of values. This was verified by:
def compare_to_scipy(xmin = 0.1, xmax = 100.0, nx = 5000, smin = 1, smax = 41, sstep = 1):
    xlist = np.linspace(xmin, xmax, nx)
    slist = np.arange(smin, smax, sstep)
    for s in slist:
        for x in xlist:
            delta = scipy.stats.chi2.sf(x, s) - chi2sf(x, s)
            if delta > 1e-10:
                print x, s, delta, scipy.stats.chi2.sf(x, s), chi2sf(x, s)

jfinkels commented on goftests/utils.py:
```diff
@@ -0,0 +1,100 @@
+    distribution. The survival function is given as:
+    1 - CDF
+    where rhe chi^2 CDF is given as
+    F(x; s) = \frac{ \gamma( x/2, s/2 ) }{ \Gamma(s/2) },
```
This should be

``` python
.. math::

   F(x; s) = ...
```

or for inline `:math:`F(x; s) = ...``.

jfinkels commented on goftests/utils.py:
```diff
@@ -0,0 +1,100 @@
+    This function matches the results from
+    scipy to numerical precision.
+    """
+    top = incomplete_gamma(x/2.0, s/2.0)
```
The `2.0`s and `1.0`s are all unnecessary as long as you have `from __future__ import division`; it's more readable to just do `x / 2`, etc.

jfinkels commented on goftests/utils.py:
```diff
@@ -0,0 +1,100 @@
+    numerical error. This function matches the results from
+    scipy to numerical precision.
+    """
+    value = 0.0
```
This doesn't need to be defined at the top of the function :). Best to leave it close to where it is actually used.

jfinkels commented on goftests/utils.py:
```diff
@@ -0,0 +1,100 @@
+        return math.gamma(s)
+    log_gamma_s = log(math.gamma(s))
+    log_x = log(x)
+    for k in range(0, 100):
```
The default first argument for range is zero, so this should just be `for k in range(100):`.

jfinkels commented on goftests/utils.py:
```diff
@@ -0,0 +1,100 @@
+    log_gamma_s = log(math.gamma(s))
+    log_x = log(x)
+    for k in range(0, 100):
+        log_num = (k + s)*log_x + (-x) + log_gamma_s
```
If the only use for `k` is in the expression `k + s`, you might want to just do `for k in range(s, s + 100):`

jfinkels commented on goftests/utils.py:
```diff
@@ -0,0 +1,100 @@
+    log_x = log(x)
+    for k in range(0, 100):
+        log_num = (k + s)*log_x + (-x) + log_gamma_s
+        log_denom = log(math.gamma(k + s + 1))
```
Can you use [`math.lgamma`](https://docs.python.org/3/library/math.html#math.lgamma) here (and on line 75)? You may have mentioned this,  but I don't remember what the explanation was. It exists in both Python 2.7 and 3.2+.

fritzo commented:
@jfinkels I believe you are now be able to merge PRs on goftests. Feel free to merge this and future PRs. I can generally only maintain on weekends. Ping me if you need to update pypi, or I can share ownership.

@tmarkovich Thanks for making this change, it's great to remove a dependency!

jfinkels commented:
@fritzo Thanks! I merged the other pull request I made, and I'll merge this one after reviewing with @tmarkovich.

tmarkovich commented:
@jfinkels Thank you for pointing out how to use math mode in doc-strings. I hope that I implemented it properly. I also took your suggestion to use lgamma. It changed no test passage but appears to be cleaner and marginally faster.

I also removed this because It makes a better message here for developers than a comment for end users:
This was implemented so that we could use the goftests module
with pypy. Given that the only scipy dependence inside goftests
was from the chi^2 survival function (1 - cdf), and then gamma
function.

@fritzo I'm glad you like it!

jfinkels reviewed (changes requested):
Hey Thomas, can you rebase this?

jfinkels commented:
This was superceded by pull request #15.
